src/GafferScene/PrimitiveVariableExists.cpp

```diff
@@ -0,0 +1,145 @@
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "IECoreScene/Primitive.h"
+
+#include "GafferScene/PrimitiveVariableExists.h"
+
+#include "Gaffer/Context.h"
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace GafferScene;
+using namespace Gaffer;
+
+IE_CORE_DEFINERUNTIMETYPED( PrimitiveVariableExists );
+
+size_t PrimitiveVariableExists::g_firstPlugIndex = 0;
+
+PrimitiveVariableExists::PrimitiveVariableExists( const std::string &name )
+	:	ComputeNode( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new ScenePlug( "in", Gaffer::Plug::In ) );
+	addChild( new StringPlug( "targetName", Gaffer::Plug::In, "P" ) );
+	addChild( new BoolPlug( "out", Gaffer::Plug::Out ) );
+}
+
+PrimitiveVariableExists::~PrimitiveVariableExists()
+{
+}
+
+ScenePlug *PrimitiveVariableExists::inPlug()
+{
+	return getChild<ScenePlug>( g_firstPlugIndex );
+}
+
+const ScenePlug *PrimitiveVariableExists::inPlug() const
+{
+	return getChild<ScenePlug>( g_firstPlugIndex );
+}
+
+StringPlug *PrimitiveVariableExists::targetNamePlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const StringPlug *PrimitiveVariableExists::targetNamePlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+BoolPlug *PrimitiveVariableExists::outPlug()
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+const BoolPlug *PrimitiveVariableExists::outPlug() const
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+void PrimitiveVariableExists::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	ComputeNode::affects( input, outputs );
+
+	if( input == inPlug()->objectPlug() || input == targetNamePlug() )
+	{
+		outputs.push_back( outPlug() );
+	}
+}
+
+void PrimitiveVariableExists::hash( const ValuePlug *output, const Context *context, IECore::MurmurHash &h ) const
+{
+	ComputeNode::hash( output, context, h );
+	if( output == outPlug() )
+	{
+		if( context->get<InternedStringVectorData>( ScenePlug::scenePathContextName, nullptr ) )
+		{
+			h.append( targetNamePlug()->hash() );
+			h.append( inPlug()->objectPlug()->hash() );
+		}
+		else
```

Comment:
If you want to use PrimitiveVariableExists to drive enabling a node, it would be pretty annoying if you had to somehow manually detect the case where you're evaluating a global plug like the setNames, and avoid querying the PrimitiveVariableExists.  ( Actually, we can't even do this with expressions currently, because expressions always depend on all their inputs - so I guess if I didn't do this here, it would require some new custom node that only queries it's input if a given context variable exists? ).